File: Array/20_valid_parentheses.cpp
class Solution {
private:
    unordered_map<char, char> map;
public:
    bool isValid(string s) {
        map[')'] = '(';
        map[']'] = '[';
        map['}'] = '{';
        stack<char> st;

        for(char c: s) {
            if(map.find(c) != map.end()) {
                if(st.empty()) return false;

                char top = st.top();
                st.pop();
                if(top != map[c]) return false; 
            } else st.push(c);
        }

        return st.empty();
    }
};

File: Stack/20_valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        hashmap = {
            ")": "(",
            "}": "{",
            "]": "["
        }

        stack = []

        for c in s:
            if c not in hashmap:
                stack.append(c)
            else:
                if not stack or stack[-1] != hashmap[c]:
                    return False
                else:
                    stack.pop()

        return True if len(stack) == 0 else False
    
# revised - 01/27/2025